### ISSUE TITLE
Performance Issues with Summing Per-Example Gradients (Caused by Fusions?)

### ISSUE THREAD

**Issue by lumip** [google/jax#1232]
I am facing an ML optimization problem which requires me to clip the per-example gradients before summing them over the minibatch and taking the optimization step. Building on the example in the README gives me access to the per-example gradients without much noticeable overhead over computing the full batch gradient (which is great!).

However, as soon as I then perform any non-trivial operation over these returned gradients, perfomance plummets: If I merely sum over the examples in the returned gradients to recover the minibatch gradient, the runtime for my first epoch increases from ~23 s to ~233 s (which includes jit compilation) and for the second from ~6 s to ~211 s.
I noticed that without summing (in which case I simply use the gradient of the first example as batch gradient), all CPU cores are continuously used (at about 70 % load). The same is true if I do not compute per-example gradients in the first place. In contrast, with the summing the load on all cores fluctuates drastically (screenshots attached).

Due to the complexity of my model, the parameters (and hence the gradients) are composed
in a tree (pytree? jaxtupletree?) that contains a total of 10 arrays, which are 1 or 2 dimensional,
so all operations on gradients require mapping over the tree. I'm not sure whether that has a significant effect on performance, but I don't believe so. The per-example gradients then add an additional leading dimension of 128 (batch_size) to each of these arrays, so the overall result of the per-example value_and_grad is a tree with arrays of shape `(128, a[, b])` and I want to sum over the first dimension in each.

Very simplified, the below is what I am currently doing
```python
    per_example_loss, per_example_grads = per_example_value_and_grad(
        loss_fn
    )(
        params, data, targets
    )

    loss_val = per_example_loss[0]

    # variant 1
    # this one is fast. basically no performance impact, all
    #   cpu cores have basically the same constant usage
    grads = jax.tree_map(lambda x: x[0], per_example_grads)

    # variant 2
    # this one causes cpu usage to vary a lot on each core and
    #   has a heavy performance impact
    grads = jax.tree_map(lambda x: np.sum(x, axis=0), per_example_grads)

    params = update_params(grads)
```
where `per_example_value_and_grad(loss_fn)` is essentially `jit(vmap(value_and_grad(loss_fn), in_axes=(None, 0, 0)))` (and I use only one of the two variants, either the summing or returning the gradient of the first example, not both).

I tried profiling as detailed in the [documentation](https://jax.readthedocs.io/en/latest/profiling.html) and noticed that there is a significant pause (of 224 s) in the generation of the output log which coincides with the starting of the observed CPU usage irregularities, so I focussed on the output generated after that (everything starting from the 12:18:16.894113 timestamp). The immediate following output is a collection of smaller operations, followed by what seems to be a summary for the encapsulating routine/expression, the major entries of which are:

```
12:18:16.896646: Execution profile for wide.jaxpr_computation__93.2970.clone.clone.clone.clone: (3.58 min @ f_nom)
12:18:16.896654:    406843241701 cycles (100.% 100Σ) ::  214579765.7 usec (         0.0 optimal) ::        2.29MFLOP/s ::        5.67kTROP/s ::      3.40MiB/s ::     0.002B/cycle :: [total]
12:18:16.896662:    207846714724 cycles (51.09% 51Σ) ::  109623793.1 usec (         0.0 optimal) ::      448.10kFLOP/s ::                    ::      1.42MiB/s ::     0.001B/cycle :: %call.13 = f32[128,400,784]{2,1,0} call(f32[128,400]{1,0} %fusion.48, f32[128,784]{1,0} %fusion.47, f32[128,784]{1,0} %fusion.45, f32[1,784]{1,0} %get-tuple-element.2622, f32[128,784]{1,0} %fusion.46, f32[1,784]{1,0} %get-tuple-element.2621, f32[128,28,28]{2,1,0} %fusion.71, u32[100352]{0} %get-tuple-element.2608, u32[2]{0} %copy.63, u32[6]{0} %get-tuple-element.2606), to_apply=%parallel_fusion.33
12:18:16.896672:    158389141742 cycles (38.93% 90Σ) ::   83538575.6 usec (         0.0 optimal) ::      559.19kFLOP/s ::                    ::      1.84MiB/s ::     0.001B/cycle :: %call.9 = f32[128,784,400]{2,1,0} call(f32[128,400]{1,0} %fusion.22, f32[128,28,28]{2,1,0} %fusion.71, u32[100352]{0} %get-tuple-element.2608, u32[2]{0} %copy.63, u32[6]{0} %get-tuple-element.2606), to_apply=%parallel_fusion.21
12:18:16.896679:     13491106112 cycles ( 3.32% 93Σ) ::    7115562.2 usec (         0.0 optimal) ::        5.60MFLOP/s ::                    ::     21.69MiB/s ::     0.012B/cycle :: %call.2 = f32[784,400]{1,0} call(f32[128,784,400]{2,1,0} %call.9, f32[] %constant.4142), to_apply=%parallel_reduce.67
12:18:16.896687:     13410656123 cycles ( 3.30% 97Σ) ::    7073130.8 usec (         0.0 optimal) ::        5.63MFLOP/s ::                    ::     21.82MiB/s ::     0.012B/cycle :: %call.3 = f32[400,784]{1,0} call(f32[128,400,784]{2,1,0} %call.13, f32[] %constant.4142), to_apply=%parallel_reduce.69
12:18:16.896695:      1216248544 cycles ( 0.30% 97Σ) ::     641481.3 usec (         0.0 optimal) ::        3.99MFLOP/s ::                    ::     15.57MiB/s ::     0.009B/cycle :: %call.4 = f32[128,400,50]{2,1,0} call(f32[128,50]{1,0} %fusion.24, f32[128,400]{1,0} %fusion.67), to_apply=%parallel_fusion.7
```

So there seems to be a massive fusion overhead (amounting to ~91% of the time being spent in only two fusion operations) and it's not entirely clear to me what's going on. From what I've gathered so far, fusioning essentially means merging low-level operations into larger "kernels" for more efficient execution (by less kernel invocation overhead), but it seems something is going terribly wrong here.

I do not have sufficient in depth understanding of jax and XLA internals to be able to tell what is going awry and how to fix or work around it. Am I somehow misusing jax? Is this a bug? If you could offer some explanation or even a hypothesis of what could be the issue here, that would be very helpful.

Unfortunately, I currently don't have data for a single iteration (only full epochs). If those would be helpful, I could provide them later. I have attached the aforementioned screenshots of CPU usage as well as the full profiling log (over two epochs) where I have annotated sections I believe are relevant, so a search for "relevant" will take you to all those locations).

Apart from that, I also noticed that if I have both variants active - with results assigned to, say, `grads` and `grads_sum` - and then don't ever use `grads_sum`, this doesn't happen. So as a side question: Does jax/XLA smartly optimize these unused variables and the corresponding operations away (or are they somehow hidden/masked in the asynchronous dispatch since the actual computation can proceed without waiting for the results)? If optimization takes place, could it be that when I'm using only the gradient of the first example it doesn't actually compute the values for the remaining ones, so that the overhead is actually in the computation of the per-example gradients after all?

### Attachments
CPU load when selecting the first of the per-example gradients. The regular dips indicate the end of an epoch:
![cpu_select_first](https://user-images.githubusercontent.com/4107081/63510173-9bfbbf80-c4e6-11e9-8a0a-bad7471d0054.jpg)

CPU load when summing the per-example gradients. Within one epoch. The portion where only one core is active marks the end of the first epoch (where a some evaluation is performed that is jit-compiled) before the next epoch starts to the right:
![cpu_sum](https://user-images.githubusercontent.com/4107081/63510178-9ef6b000-c4e6-11e9-8d39-c20be30d6a4a.jpg)

The full profiling log over two epochs with (supposedly) relevant sections marked by the "relevant" keyword.
[log_sum.txt](https://github.com/google/jax/files/3529779/log_sum.txt)

**Comment by ericmjl** [comment]
Chiming in to report that I think we are facing a similar issue with jax-unirep.

@mattjj I tried my best today to narrow down the example to a minimal reproducible example. This [gist](https://gist.github.com/ericmjl/5cf1df3570e443444e2ecceaa9a2b304) should show you what happens:

- First iteration is slow, but that's expected while we go through the first pass through the optimization.
- Next set of iterations up till iteration 32-33-ish are very fast, exactly what we expect from a JIT-compiled function.
- Past iteration 32-33-ish, iteration speed gets very, very slow.

I think this matches what @lumip uncovered up here.
